govcd/nsxt_nat_rule.go

```diff
@@ -0,0 +1,289 @@
+		return nil, err
+	}
+
+	returnObject := &NsxtNatRule{
+		NsxtNatRule:   &types.NsxtNatRule{},
+		client:        client,
+		edgeGatewayId: nsxtNat.edgeGatewayId,
+	}
+
+	err = client.OpenApiPutItem(apiVersion, urlRef, nil, natRuleConfig, returnObject.NsxtNatRule, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error updating NSX-T NAT Rule: %s", err)
+	}
+
+	return returnObject, nil
+}
+
+// Delete deletes NSX-T NAT rule
+func (nsxtNat *NsxtNatRule) Delete() error {
+	client := nsxtNat.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointNsxtNatRules
+	apiVersion, err := client.checkOpenApiEndpointCompatibility(endpoint)
+	if err != nil {
+		return err
+	}
+
+	apiVersion = elevateNsxtNatRuleApiVersion(apiVersion, client)
+
+	if nsxtNat.NsxtNatRule.ID == "" {
+		return fmt.Errorf("cannot delete NSX-T NAT rule without ID")
+	}
+
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, nsxtNat.edgeGatewayId), nsxtNat.NsxtNatRule.ID)
+	if err != nil {
+		return err
+	}
+
+	err = client.OpenApiDeleteItem(apiVersion, urlRef, nil, nil)
+
+	if err != nil {
+		return fmt.Errorf("error deleting NSX-T NAT Rule: %s", err)
+	}
+
+	return nil
+}
+
+// IsEqualTo allows to check if a rule has exactly the same fields (except ID) to the supplied rule
+// This validation is very tricky because minor version changes impact how fields are return.
+// This function relies on most common and stable fields:
+// * Name
+// * Enabled
+// * Description
+// * ExternalAddresses
+// * InternalAddresses
+// * ApplicationPortProfile.ID
+func (nsxtNat *NsxtNatRule) IsEqualTo(rule *types.NsxtNatRule) bool {
+	return natRulesEqual(nsxtNat.NsxtNatRule, rule)
+}
+
+// natRulesEqual is a helper to check if first and second supplied rules are exactly the same (except ID)
+func natRulesEqual(first, second *types.NsxtNatRule) bool {
+	util.Logger.Println("comparing NAT rule:")
+	util.Logger.Printf("%+v\n", first)
+	util.Logger.Println("against:")
+	util.Logger.Printf("%+v\n", second)
+
+	// Being an org user always returns logging as false - therefore cannot compare it.
+	// first.Logging == second.Logging &&
+
+	// These fields are returned or not returned depending on version and it is impossible to be 100% sure a minor
+	// patch does not break such comparison
+	// first.DnatExternalPort == second.DnatExternalPort &&
+	// first.SnatDestinationAddresses == second.SnatDestinationAddresses &&
+	//
+	if first.Name == second.Name &&
+		first.Enabled == second.Enabled &&
+		first.Description == second.Description &&
+
+		first.ExternalAddresses == second.ExternalAddresses &&
+		first.InternalAddresses == second.InternalAddresses &&
+		// Match either both application profiles being empty/nil, or both having the same value
+		((first.ApplicationPortProfile == second.ApplicationPortProfile) ||
+			(first.ApplicationPortProfile != nil && second.ApplicationPortProfile != nil && first.ApplicationPortProfile.ID == second.ApplicationPortProfile.ID)) {
+
```

Comment:
I think the match should be valid if one rule has `ApplicationPortProfile == nil` and the second is empty (i.e.  `ApplicationPortProfile.ID == ""`)

Reply:
Possibly. I have added two more checks to validate from both sides.